hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/index/RunIndexActionExecutor.java

```diff
@@ -0,0 +1,263 @@
+      }
+      return Option.of(indexCommitMetadata);
+    } catch (IOException e) {
+      throw new HoodieIndexException(String.format("Unable to index instant: %s", indexInstant));
+    }
+  }
+
+  private static List<HoodieInstant> getRemainingArchivedAndActiveInstantsSince(String instant, HoodieTableMetaClient metaClient) {
+    List<HoodieInstant> remainingInstantsToIndex = metaClient.getArchivedTimeline()
+        .getWriteTimeline()
+        .findInstantsAfter(instant)
+        .getInstants().collect(Collectors.toList());
+    remainingInstantsToIndex.addAll(metaClient.getActiveTimeline().getWriteTimeline().findInstantsAfter(instant).getInstants().collect(Collectors.toList()));
+    return remainingInstantsToIndex;
+  }
+
+  private static List<HoodieInstant> getCompletedArchivedAndActiveInstantsAfter(String instant, HoodieTableMetaClient metaClient) {
+    List<HoodieInstant> completedInstants = metaClient.getArchivedTimeline()
+        .filterCompletedInstants()
+        .findInstantsAfter(instant)
+        .getInstants().collect(Collectors.toList());
+    completedInstants.addAll(metaClient.getActiveTimeline().filterCompletedInstants().findInstantsAfter(instant).getInstants().collect(Collectors.toList()));
+    return completedInstants;
+  }
+
+  /**
+   * Indexing check runs for instants that completed after the base instant (in the index plan).
+   * It will check if these later instants have logged updates to metadata table or not.
+   * If not, then it will do the update. If a later instant is inflight, it will wait until it is completed or the task times out.
+   */
+  class IndexingCheckTask implements Runnable {
+
+    private final HoodieTableMetadataWriter metadataWriter;
+    private final List<HoodieInstant> instantsToIndex;
+    private final Set<String> metadataCompletedInstants;
+    private final HoodieTableMetaClient metaClient;
+
+    IndexingCheckTask(HoodieTableMetadataWriter metadataWriter,
+                      List<HoodieInstant> instantsToIndex,
+                      Set<String> metadataCompletedInstants,
+                      HoodieTableMetaClient metaClient) {
+      this.metadataWriter = metadataWriter;
+      this.instantsToIndex = instantsToIndex;
+      this.metadataCompletedInstants = metadataCompletedInstants;
+      this.metaClient = metaClient;
+    }
+
+    @Override
+    public void run() {
+      while (!Thread.interrupted()) {
+        for (HoodieInstant instant : instantsToIndex) {
+          // metadata index already updated for this instant
+          if (metadataCompletedInstants.contains(instant.getTimestamp())) {
+            currentIndexedInstant = instant.getTimestamp();
+            continue;
+          }
+          while (!instant.isCompleted()) {
+            // reload timeline and fetch instant details again wait until timeout
+            String instantTime = instant.getTimestamp();
+            Option<HoodieInstant> currentInstant = metaClient.reloadActiveTimeline()
```

Comment:
looks like we are reloading in repeated succession here. can we add a 5 sec delay between each check